Automatic regression tests

kgerheiser opened the pull request:
This is a script meant to be run in cron that will checkout UFS_UTILS, compare the hash to last build, run the regression tests if the hash has changed, and email the results.

I monitor the existence of `summary.log` to check if the tests have finished.

The driver scripts were changed so that a custom output location or project code can be set with the default value remaining the same. 

This could be the basis for automatic testing rather than just testing the develop branch (which failing code shouldn't get in to).

edwardhartnett commented:
Awesome to see this. Please cite the related issue in the description.

GeorgeGayno-NOAA commented:
@kgerheiser Is this ready to try?  What do I do?  Add ./reg_tests/rt.sh to my cron?

kgerheiser commented:
I would just try running it manually first. Set `WORK_DIR`, `MAILTO`, `PROJECT_CODE`, and `QUEUE` (or not to use the defaults). And just run `rt.sh`. It will then checkout, build, and run the regression tests.

I use a `while; sleep` loop to check if the tests have finished by looking for summary.log. I think there might be a better way to do that (ecFLow?, or depending on the jobs with `sbatch`?)

GeorgeGayno-NOAA commented:
@kgerheiser This is great.  I am testing it now.

One request: we need to account for the dev/prod switches on WCOSS.  I will need to set the crontab on both machines, but I can't run on production.  In other scripts, I have handled the prod switches with logic that reads the /etc/dev or /etc/prod files.  Then exit the script if 'prod'.  Example (ksh):

```
typeset -L1 prod_machine
prod_machine=`cat /etc/prod`
#prod_machine=`cat /etc/dev`
typeset -L1 this_machine
this_machine=`hostname -s`

if [[ $prod_machine != $this_machine ]]
then
  exit 0
fi

```

You might come up with better logic.

kgerheiser commented:
Ok, I'll do that.

kgerheiser commented:
I've added a check if running on production or develop machine.

GeorgeGayno-NOAA commented on reg_tests/rt.sh:
```diff
@@ -0,0 +1,114 @@
+
+export PROJECT_CODE=
+export QUEUE=
+
```
@kgerheiser I am getting an error at line 11.  It is looking for sorc/machine-setup.sh in my home directory.  Cron jobs on Dell default to a user's home directory.  This would work if I checked out UFS_UTILS in my home directory.  But we don't get much space in our home dir.  I think you need to move this logic to after the clone.  We cant _run_  anything on WCOSS-prod, but we can clone and build on prod.

GeorgeGayno-NOAA commented:
@kgerheiser I just tried 2cce861 on WCOSS-Dell (development).  I am getting this error:

```
+ cd ../reg_tests
+ for dir in chgres_cube grid_gen
+ cd chgres_cube
+ ./driver.wcoss_dell_p3.sh
/gpfs/dell1/stmp/George.Gayno/UFS_UTILS/reg_tests/rt.sh: line 68: ./driver.wcoss_dell_p3.sh: No such file or directory
```

And it can't seem to find 'bsub'

```
/gpfs/dell1/stmp/George.Gayno/UFS_UTILS/reg_tests/rt.sh: line 81: bsub: command not found
cat: ./driver.wcoss_dell_p3.sh: No such file or directory
```

I also tried it on WCOSS-Dell production, and your new logic to exit after the build is working.

GeorgeGayno-NOAA commented:
@kgerheiser Can we put some kind of a limit on the number of 'sleeps'? It can be something like 12 to 18 hours for the entire test suite. If it hits the limit, then exit with a message. If these tests are taking a long time to complete, there may be something wrong that should be investigated.

kgerheiser commented:
I added a timeout counter.

`driver.wcoss_dell_p3.sh` wasn't being found because I changed the name from `driver.dell.sh` in my repo and I was cloning NOAA-EMC/UFS_UTILS. I pointed it to my repo so it can be tested.

kgerheiser commented:
I have tested it on Orion, Hera, and Jet. I don't know how to use Cron on WCOSS. I think there's supposed to be a cron folder or something, but I don't have it and `crontab -e` doesn't work. You would just have to add a `module load lsf/10.1` at the top of the script to get qsub working.

GeorgeGayno-NOAA commented:
> I have tested it on Orion, Hera, and Jet. I don't know how to use Cron on WCOSS. I think there's supposed to be a cron folder or something, but I don't have it and `crontab -e` doesn't work. You would just have to add a `module load lsf/10.1` at the top of the script to get qsub working.

On Dell, you need to create a 'cron' directory off your home directory.  Then place your cron commands in a file called "mycrontab".  On Cray I was able to do 'crontab -e'

GeorgeGayno-NOAA commented:
> I have tested it on Orion, Hera, and Jet. I don't know how to use Cron on WCOSS. I think there's supposed to be a cron folder or something, but I don't have it and `crontab -e` doesn't work. You would just have to add a `module load lsf/10.1` at the top of the script to get qsub working.

There is a module purge going on (somewhere) earlier in the script.  So I had to add the lsf load before the final loop:

```
+if [[ $target == "wcoss_dell_p3" ]]; then
+    module load lsf/10.1
+fi
+
 for dir in snow2mdl global_cycle ice_blend; do
     cd $dir
```

I got the same problem with bsub not working on Cray.

GeorgeGayno-NOAA commented:
One more request: can we add a print of what node the script was started on.  For example, at the top of the script add:

`echo STARTED ON `hostname -s``

GeorgeGayno-NOAA commented:
> > I have tested it on Orion, Hera, and Jet. I don't know how to use Cron on WCOSS. I think there's supposed to be a cron folder or something, but I don't have it and `crontab -e` doesn't work. You would just have to add a `module load lsf/10.1` at the top of the script to get qsub working.
> 
> There is a module purge going on (somewhere) earlier in the script. So I had to add the lsf load before the final loop:
> 
> ```
> +if [[ $target == "wcoss_dell_p3" ]]; then
> +    module load lsf/10.1
> +fi
> +
>  for dir in snow2mdl global_cycle ice_blend; do
>      cd $dir
> ```
> 
> I got the same problem with bsub not working on Cray.

On WCOSS-Cray, this worked for me:

```
+if [[ $target == "wcoss_cray" ]]; then
+  module load xt-lsfhpc/9.1.3
+fi
+
 for dir in snow2mdl global_cycle ice_blend; do
     cd $dir

```

kgerheiser commented:
Done

kgerheiser commented:
I successfully ran it on WCOSS Dell, but on Cray snow2mdl fails for some reason.

kgerheiser commented:
I was going over the number of characters in the file name here

https://github.com/NOAA-EMC/UFS_UTILS/blob/e28b30902660e49be85e230bae912af11ed1b9ea/sorc/emcsfc_snow2mdl.fd/model_grid.F90#L89

kgerheiser commented:
I increased the length of the file name and the tests passed.

I think this is ready to be merged now. I have pointed it to use this repo instead of my fork.

GeorgeGayno-NOAA commented:
> I increased the length of the file name and the tests passed.
> 
> I think this is ready to be merged now. I have pointed it to use this repo instead of my fork.

It is working for me on all machines except Orion. The first chgres_cube test fails. On Orion, I discovered when you run from cron, you are given a very small stack size. I added a `ulimit -s 3000000` to rt.sh.  So far it is working.

kgerheiser commented:
Strange, I already have `ulimit -s unlimited` in `driver.orion.sh`. Does it need to go somewhere else?

GeorgeGayno-NOAA commented:
> Strange, I already have `ulimit -s unlimited` in `driver.orion.sh`. Does it need to go somewhere else?

Kyle, I see that.  If you look in `c96.gfs.sigio.sh` I have a note about problems setting ulimit in the driver script.  So it is odd that it is working for you.  To get the reg tests to run, I updated `rt.sh` with `ulimit -s 3000000` near the top.  Another quirk, if I try to use more than 3000000, it gets set at 3000000. That value seems to be a max value at least for me.

kgerheiser commented:
Done

GeorgeGayno-NOAA commented:
@kgerheiser It is working for me on all machines now. Let me run it via cron over the weekend as a final check. Then we can merge next week.

GeorgeGayno-NOAA commented:
@kgerheiser It worked on all five machines over the weekend. I will give my approval shortly.

GeorgeGayno-NOAA reviewed (approved):
Successfully tested from crontab on Jet, Hera, Orion, WCOSS-Cray and WCOSS-Dell over the weekend. Ready to merge.

edwardhartnett commented:
Can this be merged?

So these tests are successfully running each night? I'm not getting any emails...

kgerheiser commented:
I updated my branch with develop. 

@GeorgeGayno-NOAA if you have it setup you can add multiple emails by separating them with commas in the `MAILTO` variable.

kgerheiser commented:
I merged it. Now that it's merged you should update the script to checkout NOAA-EMC/UFS_UTILS.

GeorgeGayno-NOAA commented:
> I updated my branch with develop.
> 
> @GeorgeGayno-NOAA if you have it setup you can add multiple emails by separating them with commas in the `MAILTO` variable.

I did not know that. Should I include anyone else besides Ed and myself?

kgerheiser commented:
You can put me on it, and no spaces between the comma.
